rir/src/compiler/opt/force_dominance.cpp

```diff
@@ -578,15 +576,29 @@ bool ForceDominance::apply(Compiler&, ClosureVersion* cls, Code* code,
                             dead.insert(split);
                         break;
                     }
-                }
-            } else if (auto cast = CastType::Cast(*ip)) {
-                if (auto mk = MkArg::Cast(cast->arg<0>().val())) {
+                } else if (auto mk = MkArg::Cast(f->input()->followCasts())) {
```

Comment:
so this is
```
if (auto mkarg = MkArg::Cast(f->followCastsAndForce())) {
...
} else if (auto mk = MkArg::Cast(f->input()->followCasts())) {
...
}
```
In other words this happens *never*....

Reply:
imho this is (a) unreadable and (b) should be moved outside that if. any `force(cast*(mkarg(eager)))` should be replaced by `eager`

Reply:
the problem is that `followCastsAndForce` doesnt't stop at `MkArg` if it's eager but returns the eager value, whereas the `followCasts` returns the `MkArg`

Reply:
so the first invariant above is something that never happens with the current implementation but if that changes we would want to fix it

Reply:
I see. it is still unreadable

Reply:
also unfortunately not quite correct. if you have a downcast there it would not be correct to replace it, wouldn't it?

Reply:
imho what we want is:
```
if (auto f = Force::Cast(i)) {
  if (!f->input()->maybePromiseWrapped()) {
    f->replaceUsesWith(f->input()));
  } else {
    if (auto c = CastType::Cast(f->input())) {
      // An upcast from a value to a lazyValue is the cue that we can unwrap mkarg
      if (c->type == CastType::UpCast && !c->input()->type.maybePromiseWrapped()) {
        if (auto mk = MkArg::Cast(f->input()->followCasts())) {
          if (mk->isEager()) {
            f->replaceUsesWith(mk->eager());
          }
        }
      }
    }
  }
}
```